### PR TITLE
Groot Documentation

### PR DESCRIPTION
**Pull request by gramss**
Spin off from #87

### CODE REVIEW

File: configuration/packages/configuring-bt-navigator.rst [hunk 1]
```diff
@@ -53,6 +53,43 @@ Parameters
...
+    Consider checking out the :ref:`groot_introduction` tutorial for using this feature. 
+
+    **Attention:** Groot will only work after the behavior tree is running, which means that the nav2 stack has to receive a goal first.
+
+:groot_zmq_publisher_port:
+
+  ====== ======= ======= ========
+  Type   Default Unit    Optional
+  ------ ------- ------- --------
+  int    1666    Port    yes
```
Comment (SteveMacenski) [inline configuration/packages/configuring-bt-navigator.rst:76]: Remove unit and describe the units in the description area. Lets not change our format for parameter tables unless you'd like to update all of them for all pages ;-)

Comment (SteveMacenski) [reply configuration/packages/configuring-bt-navigator.rst:76]: and optional.

File: migration/Foxy.rst [hunk 2]
```diff
@@ -24,6 +24,14 @@ BackUp Recovery Interface Changes
 ``target.x`` in a backup recovery goal should be positive indicating the distance to drive backward.
 In both cases negative values are silently inverted.
 
+Groot Support
+*************
+
+Live Monitoring and Editing of behavior trees with Groot is now possible.
+Switching bt-xmls on the fly through a new goal request is also included.
+Check out our Groot tutorial!
```
Comment (SteveMacenski) [inline migration/Foxy.rst:32]: Link to tutorial?

File: migration/Foxy.rst [hunk 3]
```diff
@@ -24,6 +24,14 @@ BackUp Recovery Interface Changes
 ``target.x`` in a backup recovery goal should be positive indicating the distance to drive backward.
 In both cases negative values are silently inverted.
 
+Groot Support
+*************
+
+Live Monitoring and Editing of behavior trees with Groot is now possible.
+Switching bt-xmls on the fly through a new goal request is also included.
```
Comment (SteveMacenski) [inline migration/Foxy.rst:31]: This isn't part of groot, remove reference.

File: migration/Foxy.rst [hunk 4]
```diff
@@ -24,6 +24,14 @@ BackUp Recovery Interface Changes
 ``target.x`` in a backup recovery goal should be positive indicating the distance to drive backward.
 In both cases negative values are silently inverted.
 
+Groot Support
+*************
+
+Live Monitoring and Editing of behavior trees with Groot is now possible.
```
Comment (SteveMacenski) [inline migration/Foxy.rst:30]: A little more detail here would be helpful, this doesn't say alot

File: tutorials/docs/groot.rst [hunk 5]
```diff
@@ -0,0 +1,98 @@
+.. _groot_introduction:
+
+Groot - The fancy BehaviorTree Editor
```
Comment (SteveMacenski) [inline tutorials/docs/groot.rst:3]: Make this  title actually descriptive to what you're trying to show people. See other titles, be literal

File: tutorials/docs/groot.rst [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        </div>
+      </div>
+    </h1>
+
+
+
+`Groot <https://github.com/BehaviorTree/Groot>`_ is the companion application of 
+`BehaviorTree.CPP <https://github.com/BehaviorTree/BehaviorTree.CPP>`_ to create, edit, and monitor behavior trees.
+With Groot one can visualize, edit and create new behavior trees. Also, complete new custom nodes except control flow nodes can be created using Groot.
+`Recently added <https://github.com/ros-planning/navigation2/pull/1958>`_ to the ROS2 and nav2 implementation of BT.cpp V3 is live monitoring of the current status of the behavior tree.
```
Comment (SteveMacenski) [inline tutorials/docs/groot.rst:27]: Remove "recently added", this is going to be around for 10+ years. Leave no reference to time or it will be immediately outdated.

File: tutorials/docs/groot.rst [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+          <iframe width="560" height="315" src="https://www.youtube-nocookie.com/embed/Z6xCat0zaWU?autoplay=1&mute=1" frameborder="0" allowfullscreen></iframe>
+        </div>
+      </div>
+    </h1>
+
+
+
+`Groot <https://github.com/BehaviorTree/Groot>`_ is the companion application of 
+`BehaviorTree.CPP <https://github.com/BehaviorTree/BehaviorTree.CPP>`_ to create, edit, and monitor behavior trees.
+With Groot one can visualize, edit and create new behavior trees. Also, complete new custom nodes except control flow nodes can be created using Groot.
```
Comment (SteveMacenski) [inline tutorials/docs/groot.rst:26]: Why not control flow nodes? You need to add more here, you can't say something like that and run, some explanation is required.

File: tutorials/docs/groot.rst [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+Edit Custom Behavior Trees
+==========================
+
+.. .. image:: images/Groot/codecov.io_show_covered_code.png
+  :width: 80%
+
+A few steps are necessary to view and edit existing behavior trees:
+
+1. Open Groot in editor mode
+2. Select the `Load palette from file` option  (import button) near the top left corner.
```
Comment (SteveMacenski) [inline tutorials/docs/groot.rst:38]: Explain what this doing for users.

File: tutorials/docs/groot.rst [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+
+.. image:: images/Groot/groot_bt_editor.png
```
Comment (SteveMacenski) [inline tutorials/docs/groot.rst:41]: You need to have your instructions reference and explain what these figures are. You don't number them or reference them in your instructions for anyone to know what they're supposed to be showing or what steps they belong to -- or what they're trying to communicate.

File: tutorials/docs/groot.rst [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+  :width: 49%
+
+.. image:: images/Groot/groot_with_nav2_custom_nodes.png
+  :width: 49%
+
+3. Open the file `/path/to/navigation2/nav2_behavior_tree/nav2_tree_nodes.xml` to import all the behavior tree nodes used for navigation.
+4. Select `Load tree` option near the top left corner
+5. Browse the tree you want to visualize the select ok.
+
+.. image:: images/Groot/bt_w_replanning_and_recovery.png
```
Comment (SteveMacenski) [inline tutorials/docs/groot.rst:51]: Same here - these must have some kind of reference in the instructions and a description of what's being shown.

File: tutorials/docs/groot.rst [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+.. note::
+  If a tree cannot be visualized because some nodes are missing in the pallet, you might need to visit :ref:`plugins` and search for the plugin that is missing.
```
Comment (SteveMacenski) [inline tutorials/docs/groot.rst:54]: What about if someone has their own custom nodes?

File: tutorials/docs/groot.rst [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+  If you got all the optional blackboard fields from the code of the plugin, you are free to go to add the description to ``/path/to/navigation2/nav2_behavior_tree/nav2_tree_nodes.xml``.
+  Now you can reload the pallet from this updated file and should be able to work with the tree.
+
+  Monitoring mode is unaffected from this, as the whole factory with sufficient meta data gets transferred upon connection.
```
Comment (SteveMacenski) [inline tutorials/docs/groot.rst:58]: This is the editing section, don't talk about monitoring, move this to the monitoring section.

File: tutorials/docs/groot.rst [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+  This is a known issue to reduce the overhead of creating additional xml-node-description files, when the data is also available in the shared object.
+  Feel free to support adding this feature to the open source project Groot. The `feature request <https://github.com/BehaviorTree/Groot/issues/10>`_ already exists for this.
+
+Also, the Groot editor allows for creating new custom nodes via its graphical user interface. This also works with nav2 except custom control nodes.
```
Comment (SteveMacenski) [inline tutorials/docs/groot.rst:62]: This section never actaully explains how to edit the BT. Please add more here to actually show now  that you have the BT up, what do you do.

File: tutorials/docs/groot.rst [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+Live Monitoring Behavior Trees
+==============================
+
+.. raw:: html
+
+  <a href="https://gifyu.com/image/cFdR"><img src="https://s7.gifyu.com/images/Groot-Test-5FPS.gif" alt="Groot-Test-5FPS.gif" border="0" /></a>
+
+In order to live monitor a running behavior tree, a few easy steps are necessary.
+
+Behavior trees of the BT.cpp V3 kind do not automatically include monitoring.
```
Comment (SteveMacenski) [inline tutorials/docs/groot.rst:73]: The ZMQ thing is too much information. Just explain what needs to happen and the parameters needed to make it happen

File: tutorials/docs/groot.rst [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+
+  <a href="https://gifyu.com/image/cFdR"><img src="https://s7.gifyu.com/images/Groot-Test-5FPS.gif" alt="Groot-Test-5FPS.gif" border="0" /></a>
+
+In order to live monitor a running behavior tree, a few easy steps are necessary.
+
+Behavior trees of the BT.cpp V3 kind do not automatically include monitoring. 
+This is and always should be an optional feature as it involves opening an additional communication layer: ``ZMQ``.
+
+Adding the monitoring feature to nav2 therefore involves activating a special parameter listed in :ref:`configuring_bt_navigator`.
+After activating ``enable_groot_monitoring``, the zmq network ports for a ``zmq-server`` and ``zmq-publisher`` can be set optionally.
```
Comment (SteveMacenski) [inline tutorials/docs/groot.rst:77]: what is "activating" what you mean is setting it to true, so you should say that. Be explicit, this is the official instructions for the feature, don't skip over details.

File: tutorials/docs/groot.rst [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+
+Given the launch-parameters are set correctly, the navigation2 stack has to be started normally first.
+Regardless of simulation or real hardware, **monitoring only works if the behavior tree is currently running!**
+
+A step-by-step guide for simulation can look like this:
+
+  1. complete :ref:`getting_started` and be able to run the tb3 simulation
```
Comment (SteveMacenski) [inline tutorials/docs/groot.rst:84]: Capitalize all first words in bullets

File: tutorials/docs/groot.rst [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+  If you got all the optional blackboard fields from the code of the plugin, you are free to go to add the description to ``/path/to/navigation2/nav2_behavior_tree/nav2_tree_nodes.xml``.
+  Now you can reload the pallet from this updated file and should be able to work with the tree.
+
+  Monitoring mode is unaffected from this, as the whole factory with sufficient meta data gets transferred upon connection.
+  This is a known issue to reduce the overhead of creating additional xml-node-description files, when the data is also available in the shared object.
+  Feel free to support adding this feature to the open source project Groot. The `feature request <https://github.com/BehaviorTree/Groot/issues/10>`_ already exists for this.
+
+Also, the Groot editor allows for creating new custom nodes via its graphical user interface. This also works with nav2 except custom control nodes.
+
+Live Monitoring Behavior Trees
```
Comment (SteveMacenski) [inline tutorials/docs/groot.rst:64]: You never introduce what this means. Have a introduction paragraph letting people know what you're talking about and enabling.

File: tutorials/docs/groot.rst [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+After activating ``enable_groot_monitoring``, the zmq network ports for a ``zmq-server`` and ``zmq-publisher`` can be set optionally.
+
+Given the launch-parameters are set correctly, the navigation2 stack has to be started normally first.
+Regardless of simulation or real hardware, **monitoring only works if the behavior tree is currently running!**
+
+A step-by-step guide for simulation can look like this:
+
+  1. complete :ref:`getting_started` and be able to run the tb3 simulation
+  2. set ``enable_groot_monitoring`` to ``True`` in the ``params.yaml`` file
+  3. (re-)start the tb3 simulation (optionally with ``headless:=True`` for only RVIZ as GUI)
```
Comment (SteveMacenski) [inline tutorials/docs/groot.rst:86]: Remove all things in parentheses. The additional commentary is too much and it makes it hard to read

File: tutorials/docs/groot.rst (hunk 15 above)
Comment (SteveMacenski) [reply tutorials/docs/groot.rst:77]: Is the zmq publisher/server details so important you should talk about them in the first few lines of the section? Perhaps this should be moved to the last thing to mention that you can change these ports optionally but the defaults match the defaults in groot.

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
My recommendation is to slow down and think like someone that doesn't know anything about this. It skips over details and actually introducing the concepts you're explaining how to enable. At the point someone clicks on this article, they might not even know what groot is, so explain it to them. They probably have only a cursory understanding of a BT, so show how to edit it and the structure of it.

**Comment by SteveMacenski** [comment]
@gramss what's the story here?

File: configuration/packages/configuring-bt-navigator.rst (hunk 1 above)
Comment (gramss) [reply configuration/packages/configuring-bt-navigator.rst:76]: this is now solved and already merged. Should be removed from this PR now..

Comment (gramss) [reply configuration/packages/configuring-bt-navigator.rst:76]: locally done (as rebased)

File: tutorials/docs/groot.rst (hunk 5 above)
Comment (gramss) [reply tutorials/docs/groot.rst:3]: It's the official subtitle of Groot.. :D But yes. Something more literal is more suitable

File: tutorials/docs/groot.rst (hunk 6 above)
Comment (gramss) [reply tutorials/docs/groot.rst:27]: done

File: tutorials/docs/groot.rst (hunk 7 above)
Comment (gramss) [reply tutorials/docs/groot.rst:26]: I'll add a section for creating Custom Nodes with Groot. I do not see any reason why it does not work. I had this information from a readme in nav2 about groot.. Probably outdated.

I tested it and it looks the same as the ones we already have in our big nodes.xml
So no reason why this would not work..

File: tutorials/docs/groot.rst (hunk 18 above)
Comment (gramss) [reply tutorials/docs/groot.rst]: done

File: tutorials/docs/groot.rst (hunk 17 above)
Comment (gramss) [reply tutorials/docs/groot.rst:145]: I added one

File: tutorials/docs/groot.rst (hunk 16 above)
Comment (gramss) [reply tutorials/docs/groot.rst]: done

File: tutorials/docs/groot.rst (hunk 15 above)
Comment (gramss) [reply tutorials/docs/groot.rst]: I hope I have done it to your understanding

File: tutorials/docs/groot.rst (hunk 8 above)
Comment (gramss) [reply tutorials/docs/groot.rst]: done

File: tutorials/docs/groot.rst (hunk 9 above)
Comment (gramss) [reply tutorials/docs/groot.rst]: Took me a while but I think I figured it out how to use figures correctly in rst. I did not find another use of this in the current docs

File: tutorials/docs/groot.rst (hunk 11 above)
Comment (gramss) [reply tutorials/docs/groot.rst]: I tried to describe this in detail now + altered the section of interest here

File: tutorials/docs/groot.rst (hunk 13 above)
Comment (gramss) [reply tutorials/docs/groot.rst]: This is still missing..

**Comment by gramss** [comment]
Okay.

So I went through all your comments and created a new, more detailed draft.
I just noticed that a section to edit the BT is still missing. This would be good to have it also there.

I'm unsure if this is worthy a gif or just some text describing how to properly interact with the editing interface of Groot.

I'm eager to listen what you think of this new draft.

**Comment by SteveMacenski** [comment]
> I'm unsure if this is worthy a gif or just some text describing how to properly interact with the editing interface of Groot.

Gifs and images are always good!

>  I just noticed that a section to edit the BT is still missing. This would be good to have it also there.

Agreed, that's a necessary element for a full look at Groot

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
I stopped reading about half way through, this needs alot of work.

Please, please, please **slow down**. You fail to do alot of basics and just jump into things without ever explaining or introducing them (but then talk at length of things that aren't relevant for groot).

An introduction section should start the document. In this section, we should introduce what we're going to do, why we're doing it, and the tools used to do it. 

Every subsequent section should at **least** have an introduction paragraph explaining what this section is going to do and examples of why we do this. You just jump right into things with no context or explanation. 

Every instruction should have a detailed explanation of what you're doing and why. If you load a palette, the least you should do is explain what a palette is and what it is used for. These are not obvious things.

I think it would help you alot to walk away from this for a few days and then come back with fresh eyes. Imagine you're writing this document for someone that knows nothing about groot, palettes, editing BT files, etc. Introduce it like you'd want someone to explain a new technology to you. 

In other words, slow down and be complete

**Comment by SteveMacenski** [comment]
Working on bringing this to the documentation today. Closing and will reopen new PR once complete

**Comment by SteveMacenski** [comment]
https://github.com/ros-planning/navigation.ros.org/pull/238
